ehancements to connection slot mitigation

ptschip opened the pull request:
Added the ability to screen heavy hitters and also eviction candidates.   Heavy hitters are nodes that come in and make repeated connections within a short period (we do have the right now).  60 seconds is used in this case and any nodes doing more than 4 connections in the timeframe will get a 1 hour ban.  Similarly we have nodes that are not useful and get evicted frequently...these are nodes that typically just listen to INV 's.  If we get more than 15 to 30 evictions in a 30 minute period then we give them a 4 hour ban (we also have some of these right now).  Both these cases are far enough out to the extremes that the bitnodes crawlers remain unaffected.  

The tracking map and struct forms a foundations to a more robuts DOSmanager.  I"d like to eventually put everything in a DOSman.cpp  where we can track all kinds of node activity and persist it between startups.

Commit:
Set a 60 minute ban for heavy hitters that connect/disconnect too frequently

If a node makes more than 4 connections in a one minute period then give
them a 60 minute ban.

Then inbound connection counts are decayed over one minute so it could
be 4 connections in rapid succession or 8 spread out over a one minute
period that trigger the ban.

Commit:
Adding Eviction tracking for each IP address.

Building on Connection tracking, Eviction tracking will ban
a peer for 4 hours if they get evicted more than between 15
and 30 times within a 30 minute period depending on how closely
spaced the interval of eviction is (Similar to the Connection
tracking the eviction counts are gradually decayed over a 30
minute period).

Commit:
Gradually decay nActivityBytes over a 2 hour period

This slowly deprioritizes nodes that have gone stale while allowing
more active newer nodes to climb the ladder of priority.

Commit:
Choke off txn inventory messages

If nActivityBytes is zero and the peer has been connected for more
than two minutes then do not send them INV traffic for transations.
However, we continue to send INV for block announcements.

This prevents bad actors from connecting and just listening to
INV traffic for the purpose of data gathering.

Commit:
Trivial Code simplification and logging changes

ptschip commented:
taking this out of WIP.  The next step up will require building a more comprehensive dos manager.  A separate DOSman.cpp would be a good idea where we can put everything and track more peer data and also track the between startups but that's beyond the scope of this PR.

Commit:
Stop deprioritizing connections based on netgroups

There is no need to do this anymore now that we have deprioritization
based on the usefulness or activity of connected nodes.  And furthermore
it can cause unintended consequences in the setup of an Xpedited or
test network which uses addnode and has more than one connection coming
in from the same network group.

gandrewstone commented:
is this ready for merge?

Justaphf commented on src/net.cpp:
```diff
@@ -1090,6 +1133,35 @@ static void AcceptConnection(const ListenSocket& hListenSocket) {
+            int nHoursToBan = 4;
+            CNode::Ban((CNetAddr)addr, BanReasonNodeMisbehaving, nHoursToBan*60*60);
+            LogPrintf("Banning %s for %d hours: Too many connection attempts - connection dropped\n", addr.ToString(), nHoursToBan);
+            return;
```
I think you need to call `CloseSocket(hSocket)` here since the socket is open but because of the ban we are returning.

ptschip replied:
I was letting the ban take effect on the next connection that comes in
becuase we don't want to disconnect whitelisted peers, but perhaps the
code is not entirely clear.  I should add that in and then we also don't
want to print out the log statement if the peer is whitelisted..so I'll
make that change, thanks

On 02/09/2016 9:26 PM, Justaphf wrote:

> In src/net.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62#discussion_r77431155:
> 
> > -            double nRatioElapsed = (double)nTimeElapsed / 60;
> > -            nConnections = mapInboundConnectionTracker[ipAddress].nConnections - (nRatioElapsed \* mapInboundConnectionTracker[ipAddress].nConnections);
> > -            if (nConnections < 0) 
> > -                nConnections = 0;
> > -        }
> >   +
> > -        nConnections += 1;
> > -        mapInboundConnectionTracker[ipAddress].nConnections = nConnections;
> > -        mapInboundConnectionTracker[ipAddress].nLastConnectionTime = GetTime();
> >   +
> > -        LogPrint("evict", "Number of Connection attempts is %f for %s\n", nConnections, addr.ToString());
> > -        if (nConnections > 4) {
> > -            int nHoursToBan = 4;
> > -            CNode::Ban((CNetAddr)addr, BanReasonNodeMisbehaving, nHoursToBan_60_60);
> > -            LogPrintf("Banning %s for %d hours: Too many connection attempts - connection dropped\n", addr.ToString(), nHoursToBan);
> > -            return;
> 
> I think you need to call |CloseSocket(hSocket)| here since the socket
> is open but because of the ban we are returning.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62/files/2192d2c5b6cda21ed24bd4a5928d88893087ac71#r77431155,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0MG6oosPhkWw0MepteN05D5aGY5mks5qmPbagaJpZM4Jc23I.

ptschip commented:
Yes, it's been well tested...

On 02/09/2016 6:53 PM, gandrewstone wrote:
>
> is this ready for merge?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62#issuecomment-244520121>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0BKEVmi-x1J0gggHGNUiHgipeD9rks5qmNMFgaJpZM4Jc23I>.
>

Commit:
Explicitly disconnect if too many connection attempts

When we get more than the limit of connection attempts in 1
minute then disconnect the peer right away instead of waiting
for the ban logic to do it.  Also, don't ban whitelisted peers.

ptschip commented:
ok, i just pushed up the last changed that @justaphf recommended

On 02/09/2016 6:53 PM, gandrewstone wrote:
>
> is this ready for merge?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62#issuecomment-244520121>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0BKEVmi-x1J0gggHGNUiHgipeD9rks5qmNMFgaJpZM4Jc23I>.
>

gandrewstone commented:
thanks for this!!  Personally, I get a lot of spam-like connections...
